{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fbd9b960",
   "metadata": {},
   "source": [
    "## Ovrlpy can help identify spurious spatial cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4428d124",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "import anndata as ad\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import ovrlpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3548f02b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib_scalebar.scalebar import ScaleBar\n",
    "from matplotlib.patches import Rectangle\n",
    "from typing import Any\n",
    "import matplotlib.lines as mlines\n",
    "\n",
    "SCALEBAR_PARAMS: dict[str, Any] = {\n",
    "    \"dx\": 1,\n",
    "    \"units\": \"um\",\n",
    "    \"length_fraction\": 0.1,\n",
    "    \"location\": \"lower right\",\n",
    "    \"pad\": 0.1,\n",
    "    \"frameon\": False\n",
    "}\n",
    "\n",
    "def _plot_scalebar(ax, dx=1, units=\"um\", fontsize=5, color=\"black\", box_alpha=0, **kwargs):\n",
    "    scalebar = ScaleBar(\n",
    "        dx,\n",
    "        units=units,\n",
    "        scale_loc=\"top\", \n",
    "        sep=1,\n",
    "        **kwargs\n",
    "    )\n",
    "    scalebar.set_font_properties({\"size\": fontsize})\n",
    "    scalebar.linewidth = 0.3\n",
    "    scalebar.box_alpha = box_alpha\n",
    "    scalebar.color = color\n",
    "    ax.add_artist(scalebar)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e6cd6d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "CM = 1/2.54\n",
    "\n",
    "fig_dir = Path(\"./fig/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1c27a52",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "\n",
    "BIH_CMAP = LinearSegmentedColormap.from_list(\n",
    "    \"BIH\",\n",
    "    [\n",
    "        \"#430541\",\n",
    "        \"mediumvioletred\",\n",
    "        \"violet\",\n",
    "        \"powderblue\",\n",
    "        \"powderblue\",\n",
    "        \"white\",\n",
    "        \"white\",\n",
    "    ][::-1],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "817287e4",
   "metadata": {},
   "source": [
    "### data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9006155",
   "metadata": {},
   "source": [
    "#### Transcripts in the Tissue Section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a10dc4b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "MERFISH_data_folder_path = Path(\"../data/mouse_hypothalamus/MERFISH/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5717f0f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = [\n",
    "        \"Centroid_X\",\n",
    "        \"Centroid_Y\",\n",
    "        \"Centroid_Z\",\n",
    "        \"Gene_name\",\n",
    "]\n",
    "\n",
    "signal_coordinate_df = pd.read_csv(MERFISH_data_folder_path/\"merfish_barcodes_example.csv\", usecols=columns).rename(\n",
    "    columns={\n",
    "        \"Centroid_X\": \"x\",\n",
    "        \"Centroid_Y\": \"y\",\n",
    "        \"Centroid_Z\": \"z\",\n",
    "        \"Gene_name\": \"gene\",\n",
    "    }\n",
    ")\n",
    "\n",
    "signal_coordinate_df = signal_coordinate_df.loc[~signal_coordinate_df[\"gene\"].str.contains(\"Blank|NegControl\")].copy()\n",
    "\n",
    "signal_coordinate_df[\"gene\"] = signal_coordinate_df[\"gene\"].astype(\"category\")\n",
    "\n",
    "coordinate_x_m = signal_coordinate_df[\"x\"].min()\n",
    "coordinate_y_m = signal_coordinate_df[\"y\"].min()\n",
    "signal_coordinate_df[\"x\"] = signal_coordinate_df[\"x\"] - coordinate_x_m\n",
    "signal_coordinate_df[\"y\"] = signal_coordinate_df[\"y\"] - coordinate_y_m\n",
    "\n",
    "signal_coordinate_df = signal_coordinate_df.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6b1d394",
   "metadata": {},
   "source": [
    "#### Results of BANKSY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9ee29ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "banksy_folder_path = Path(\"../data/banksy_results/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e738883",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = [\"Centroid_X\", \"Centroid_Y\", \"Bregma\", \"lam0.2\"]\n",
    "\n",
    "banksy_result = pd.read_csv(\n",
    "    banksy_folder_path/\"banksy_cluster.txt\",\n",
    "    usecols=columns,\n",
    "    sep=\"\\t\"\n",
    ").rename(\n",
    "    columns={\n",
    "        \"Centroid_X\": \"x\",\n",
    "        \"Centroid_Y\": \"y\",\n",
    "        \"Bregma\": \"Bregma\",\n",
    "        \"lam0.2\": \"banksy_cluster\",\n",
    "    }\n",
    ")\n",
    "\n",
    "banksy_result = banksy_result[banksy_result[\"Bregma\"] == -0.24]\n",
    "\n",
    "banksy_result[\"x\"] = banksy_result[\"x\"] - coordinate_x_m\n",
    "banksy_result[\"y\"] = banksy_result[\"y\"] - coordinate_y_m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4edd6b1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "MOD1_df = banksy_result[banksy_result['banksy_cluster'] == 8]\n",
    "MOD2_df = banksy_result[banksy_result['banksy_cluster'] == 7]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18518658",
   "metadata": {},
   "source": [
    "differentially expressed genes identified by BANKSY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4866f3ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# all differentially expressed genes\n",
    "DE_genes = ['Mlc1', 'Dgkk', 'Cbln2', 'Syt4', 'Gad1', 'Plin3', 'Gnrh1', 'Sln', 'Gjc3', 'Mbp', 'Lpar1', 'Trh', 'Ucn3', 'Cck']\n",
    "# DE_genes_MOD2: 7\n",
    "DE_genes_MOD2 = ['Mlc1', 'Dgkk', 'Cbln2', 'Syt4', 'Gad1', 'Plin3', 'Gnrh1', 'Sln', 'Gjc3']\n",
    "# DE_genes_MOD1: 8\n",
    "DE_genes_MOD1 = ['Mbp', 'Lpar1', 'Trh', 'Ucn3', 'Cck']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae8d9d24",
   "metadata": {},
   "outputs": [],
   "source": [
    "MOD1_signals = signal_coordinate_df[signal_coordinate_df['gene'].isin(DE_genes_MOD1)].copy()\n",
    "MOD2_signals = signal_coordinate_df[signal_coordinate_df['gene'].isin(DE_genes_MOD2)].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a04e7637",
   "metadata": {},
   "source": [
    "#### Segmentation Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0156ccaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "merfish_data = pd.read_csv(MERFISH_data_folder_path / \"merfish_all_cells.csv\").rename(columns={\"Centroid_X\": \"x\", \"Centroid_Y\": \"y\"})\n",
    "\n",
    "merfish_data = merfish_data.drop(columns=[col for col in merfish_data.columns if col == 'Fos' or col.startswith('Blank_')])\n",
    "merfish_data = merfish_data[merfish_data[\"Cell_class\"] != \"Ambiguous\"]\n",
    "merfish_data = merfish_data[merfish_data[\"Animal_ID\"] == 1]\n",
    "merfish_data = merfish_data[merfish_data[\"Bregma\"] == -0.24]\n",
    "\n",
    "merfish_data[\"x\"] -= coordinate_x_m\n",
    "merfish_data[\"y\"] -= coordinate_y_m\n",
    "\n",
    "merfish_data = merfish_data.merge(\n",
    "    banksy_result[[\"x\", \"y\", \"banksy_cluster\"]],\n",
    "    on=[\"x\", \"y\"],\n",
    "    how=\"left\"\n",
    ")\n",
    "merfish_data = merfish_data.rename(columns={\"banksy_cluster\": \"banksy\"})\n",
    "\n",
    "cell_class_m = {\n",
    "    'Astrocyte': 'Astrocyte',\n",
    "    'Endothelial 1': 'Endothelial',\n",
    "    'Endothelial 2': 'Endothelial',\n",
    "    'Endothelial 3': 'Endothelial',\n",
    "    'Ependymal': 'Ependymal',\n",
    "    'Excitatory': 'Excitatory',\n",
    "    'Inhibitory': 'Inhibitory',\n",
    "    'Microglia': 'Microglia',\n",
    "    'OD Immature 1': 'OD Immature',\n",
    "    'OD Immature 2': 'OD Immature',\n",
    "    'OD Mature 1': 'OD Mature',\n",
    "    'OD Mature 2': 'OD Mature',\n",
    "    'OD Mature 3': 'OD Mature',\n",
    "    'OD Mature 4': 'OD Mature',\n",
    "    'Pericytes': 'Pericytes'\n",
    "}\n",
    "\n",
    "merfish_data[\"Cell_class\"] = merfish_data[\"Cell_class\"].map(cell_class_m).fillna(\"Other\")\n",
    "\n",
    "merfish_data = merfish_data.sort_values(by=\"Cell_class\").copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eba4151a",
   "metadata": {},
   "source": [
    "#### Cell boundaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5eb4ba52",
   "metadata": {},
   "outputs": [],
   "source": [
    "boundaries_df = pd.read_csv(MERFISH_data_folder_path/'cellboundaries_example_animal.csv')\n",
    "boundaries_df = boundaries_df.dropna(subset=[\"boundaryX\", \"boundaryY\"])\n",
    "\n",
    "cell_ids = merfish_data[\"Cell_ID\"]\n",
    "boundaries_df = boundaries_df[boundaries_df[\"feature_uID\"].isin(cell_ids)]\n",
    "boundaries_df = boundaries_df.merge(\n",
    "    merfish_data[[\"Cell_ID\", \"x\", \"y\", \"banksy\", \"Cell_class\"]],\n",
    "    left_on=\"feature_uID\",\n",
    "    right_on=\"Cell_ID\",\n",
    "    how=\"inner\"\n",
    ").drop(columns=[\"Cell_ID\"])\n",
    "\n",
    "boundaries_df[\"boundaryX\"] = boundaries_df[\"boundaryX\"].apply(lambda x: [float(i) for i in x.split(\";\")] if isinstance(x, str) else x)\n",
    "boundaries_df[\"boundaryY\"] = boundaries_df[\"boundaryY\"].apply(lambda x: [float(i) for i in x.split(\";\")] if isinstance(x, str) else x)\n",
    "\n",
    "boundaries_df[\"boundaryX\"] = boundaries_df[\"boundaryX\"].apply(lambda x: [i - coordinate_x_m for i in x] if isinstance(x, list) else x)\n",
    "boundaries_df[\"boundaryY\"] = boundaries_df[\"boundaryY\"].apply(lambda x: [i - coordinate_y_m for i in x] if isinstance(x, list) else x)\n",
    "\n",
    "boundaries_df = boundaries_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1c0ccdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "MOD1_boundaries = boundaries_df[boundaries_df['banksy'] == 8].copy()\n",
    "MOD2_boundaries = boundaries_df[boundaries_df['banksy'] == 7].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28cc1222",
   "metadata": {},
   "source": [
    "#### Matched Single-Cell RNA Sequencing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a37584c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import sparse\n",
    "from scipy.io import mmread\n",
    "from sklearn.preprocessing import MaxAbsScaler\n",
    "from pandas.api.types import CategoricalDtype\n",
    "\n",
    "def load_scRNA_data(mtx_path, barcodes_path, genes_path, meta_path, cell_class_filter, neuron_cluster = False):\n",
    "    X = mmread(mtx_path).tocsr()\n",
    "    cell_ids = pd.read_csv(barcodes_path, sep=\"\\t\", header=None)[0].values\n",
    "    gene_names = pd.read_csv(genes_path, sep=\"\\t\", header=None)[1].values\n",
    "\n",
    "    adata = ad.AnnData(X=X.T)\n",
    "    adata.var_names = gene_names\n",
    "    adata.obs_names = cell_ids\n",
    "    adata.var_names_make_unique()\n",
    "\n",
    "    meta = pd.read_excel(meta_path).rename(columns={\n",
    "        \"Cell name\": \"Cell_name\",\n",
    "        \"Sex\": \"Sex\",\n",
    "        \"Replicate number\": \"Rep\",\n",
    "        \"Cell class (determined from clustering of all cells)\": \"Cell_class\",\n",
    "        \"Non-neuronal cluster (determined from clustering of all cells)\": \"Non_neuronal_cluster\",\n",
    "        \"Neuronal cluster (determined from clustering of inhibitory or excitatory neurons)\": \"Neuronal_cluster\"\n",
    "    })\n",
    "    meta = meta.set_index(\"Cell_name\")\n",
    "    meta = meta.loc[meta[\"Cell_class\"].isin(cell_class_filter.keys())]\n",
    "    adata = adata[adata.obs_names.isin(meta.index)].copy()\n",
    "    if neuron_cluster:\n",
    "        adata.obs = meta.loc[adata.obs_names, [\"Cell_class\", \"Neuronal_cluster\"]].copy()\n",
    "    else:\n",
    "        adata.obs = meta.loc[adata.obs_names, [\"Cell_class\"]].copy()\n",
    "    adata.obs[\"Cell_class\"] = adata.obs[\"Cell_class\"].map(cell_class_filter)\n",
    "\n",
    "    cell_class_order = list(cell_class_filter.values())\n",
    "    cat_dtype = CategoricalDtype(categories=cell_class_order, ordered=True)\n",
    "    adata.obs[\"Cell_class\"] = adata.obs[\"Cell_class\"].astype(cat_dtype)\n",
    "\n",
    "    mt_mask = adata.var_names.str.startswith(\"mt\")\n",
    "    mt_fraction = np.array(adata[:, mt_mask].X.sum(axis=1)).flatten() / (np.array(adata.X.sum(axis=1)).flatten() + 1e-6)\n",
    "    adata = adata[mt_fraction < 0.2, :].copy()\n",
    "\n",
    "    nonzero_counts = np.array((adata.X != 0).sum(axis=1)).flatten()\n",
    "    adata = adata[nonzero_counts > 1000, :].copy()\n",
    "\n",
    "    blank_mask = ~adata.var_names.str.startswith(\"Blank\")\n",
    "    adata = adata[:, blank_mask].copy()\n",
    "\n",
    "    sc_total = np.array(adata.X.sum(axis=1)).flatten() + 1e-6\n",
    "    normalizer = sparse.diags(10_000 / sc_total)\n",
    "    adata.X = normalizer.dot(adata.X)\n",
    "\n",
    "    adata.X = adata.X.log1p()\n",
    "\n",
    "    scaler = MaxAbsScaler()\n",
    "    adata.X = scaler.fit_transform(adata.X)\n",
    "\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6a925ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = Path(\"../data/mouse_hypothalamus/SingleCell/\")\n",
    "\n",
    "cell_class_dict = {\n",
    "    'Astrocytes': 'Astrocytes',\n",
    "    'Microglia': 'Microglia',\n",
    "    'Macrophage': 'Macrophage',\n",
    "    'Excitatory': 'Excitatory',\n",
    "    'Inhibitory': 'Inhibitory',\n",
    "    'Immature oligodendrocyte': 'OD immature',\n",
    "    'Newly formed oligodendrocyte': 'OD newly formed',\n",
    "    'Mature oligodendrocyte': 'OD mature',\n",
    "    'Ependymal': 'Ependymal',\n",
    "    'Endothelial': 'Endothelial',\n",
    "    'Mural': 'Mural',\n",
    "    'Fibroblast': 'Fibroblast',\n",
    "}\n",
    "\n",
    "adata = load_scRNA_data(\n",
    "    mtx_path = data_path / \"GSE113576_matrix.mtx\",\n",
    "    barcodes_path = data_path / \"GSE113576_barcodes.tsv\",\n",
    "    genes_path = data_path / \"GSE113576_genes.tsv\",\n",
    "    meta_path = data_path / \"aau5324_Moffitt_Table-S1.xlsx\",\n",
    "    cell_class_filter = cell_class_dict\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7d645f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[adata.obs.sort_values(\"Cell_class\").index]\n",
    "\n",
    "sc_data = adata.to_df()\n",
    "sc_DE_MOD1_df = sc_data[DE_genes_MOD1]\n",
    "sc_DE_MOD2_df = sc_data[DE_genes_MOD2]\n",
    "sc_DE = pd.concat([sc_DE_MOD1_df, sc_DE_MOD2_df], axis=1)\n",
    "\n",
    "sc_cell_class = adata.obs['Cell_class']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d8a309b",
   "metadata": {},
   "source": [
    "### Ovrlpy pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "054e6fb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_components = 15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "724be437",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = ovrlpy.Ovrlp(\n",
    "    signal_coordinate_df,\n",
    "    n_components=n_components,\n",
    "    n_workers=8,\n",
    ")\n",
    "\n",
    "dataset.analyse()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db6d985b",
   "metadata": {},
   "outputs": [],
   "source": [
    "signal_integrity = dataset.integrity_map\n",
    "signal_strength = dataset.signal_map\n",
    "pseudocells = dataset.pseudocells\n",
    "\n",
    "doublets = dataset.detect_doublets(min_signal=3, integrity_sigma=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a288ac8",
   "metadata": {},
   "source": [
    "### cell type map\n",
    "Cell types are colored according to the original publication ([Moffitt et al., 2018](https://www.science.org/doi/10.1126/science.aau5324))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "127ac196",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_labels = sorted(merfish_data['Cell_class'].unique())\n",
    "colormap = plt.cm.get_cmap('tab20b', len(all_labels))\n",
    "label_to_color = {label: colormap(i) for i, label in enumerate(all_labels)}\n",
    "colors = merfish_data['Cell_class'].map(label_to_color)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6*CM, 6*CM), dpi=600)\n",
    "\n",
    "ax.scatter(\n",
    "    merfish_data[\"x\"], merfish_data[\"y\"],\n",
    "    s=2, c=colors, edgecolors='none'\n",
    ")\n",
    "\n",
    "ax.set(\n",
    "    xticks=[], yticks=[],\n",
    "    xlim=(-10, 1810), ylim=(-10, 1810),\n",
    "    aspect='equal'\n",
    ")\n",
    "\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_linewidth(0.3)\n",
    "\n",
    "plt.title('mouse hypothalamus', fontsize=6, fontweight='light')\n",
    "\n",
    "handles = [\n",
    "    mlines.Line2D(\n",
    "        [0], [0], marker='.', linestyle='None',\n",
    "        color='none',\n",
    "        markerfacecolor=label_to_color[label],\n",
    "        markeredgecolor='none',\n",
    "        markersize=4,\n",
    "        label=label\n",
    "    )\n",
    "    for label in all_labels\n",
    "]\n",
    "\n",
    "ax.legend(\n",
    "    handles=handles, fontsize=4, loc='upper center',\n",
    "    frameon=False, bbox_to_anchor=(0.5, 0), ncol=3\n",
    ")\n",
    "\n",
    "plt.subplots_adjust(right=0.8)\n",
    "plt.tight_layout()\n",
    "fig.savefig(fig_dir / \"CellTypes.pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "265717a5",
   "metadata": {},
   "source": [
    "### MOD1 marker transcripts\n",
    "`BANKSY` ([Singhal, et al.](https://www.nature.com/articles/s41588-024-01664-3)) detects two subtypes of mature oligodendrocytes (MOD1 and MOD2) and their subtype-specific markers.   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e7ccb5f",
   "metadata": {},
   "source": [
    "#### MOD1 marker transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e44b01c",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(6*CM, 6*CM), dpi=600)\n",
    "\n",
    "plt.scatter(\n",
    "    MOD1_signals[\"x\"],\n",
    "    MOD1_signals[\"y\"],\n",
    "    s=0.15,\n",
    "    c='salmon', \n",
    "    edgecolors='none',\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    MOD1_df[\"x\"],\n",
    "    MOD1_df[\"y\"],\n",
    "    s=0.5, \n",
    "    facecolors='none',\n",
    "    edgecolors=\"#080808\",\n",
    "    linewidths=0.13\n",
    ")\n",
    "\n",
    "label_to_color = {\n",
    "    'MOD1 Marker Transcripts': 'salmon',\n",
    "    'MOD1 Cells': '#080808',\n",
    "}\n",
    "\n",
    "handles = [\n",
    "    mlines.Line2D(\n",
    "        [0], [0],\n",
    "        marker='o',\n",
    "        color='none',\n",
    "        markerfacecolor=color if 'Marker' in label else 'none',\n",
    "        markeredgecolor=color,\n",
    "        linestyle='None',\n",
    "        markersize=2,\n",
    "        markeredgewidth=0.3 if 'Marker' not in label else 0,\n",
    "        label=label\n",
    "    )\n",
    "    for label, color in label_to_color.items()\n",
    "]\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "ax.legend(handles=handles, fontsize=4, loc='upper center', frameon=False, \n",
    "          bbox_to_anchor=(0.5, 0), ncol=2)\n",
    "\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "ax.spines[[\"top\", \"right\", \"left\", \"bottom\"]].set_linewidth(0.3)\n",
    "ax.set_aspect('equal')\n",
    "ax.set_xlim(-10,1810)\n",
    "ax.set_ylim(-10,1810)\n",
    "plt.title(\"MOD1 cells & marker transcripts\", fontsize=6)\n",
    "plt.tight_layout()  \n",
    "plt.savefig(fig_dir /\"MOD1marker.pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fc0f591",
   "metadata": {},
   "source": [
    "#### MOD2 marker transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed71d281",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(6*CM, 6*CM), dpi=600)\n",
    "\n",
    "plt.scatter(\n",
    "    MOD2_signals[\"x\"],\n",
    "    MOD2_signals[\"y\"],\n",
    "    s=0.15,\n",
    "    c='lightblue', \n",
    "    edgecolors='none'\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    MOD2_df[\"x\"],\n",
    "    MOD2_df[\"y\"],\n",
    "    s=0.5, \n",
    "    facecolors='none',\n",
    "    edgecolors=\"#080808\",\n",
    "    linewidths=0.13\n",
    ")\n",
    "\n",
    "label_to_color = {\n",
    "    'MOD2 Marker Transcripts': 'lightblue',\n",
    "    'MOD2 Cells': '#080808',\n",
    "}\n",
    "\n",
    "handles = [\n",
    "    mlines.Line2D(\n",
    "        [0], [0],\n",
    "        marker='o',\n",
    "        color='none',\n",
    "        markerfacecolor=color if 'Marker' in label else 'none',\n",
    "        markeredgecolor=color,\n",
    "        linestyle='None',\n",
    "        markersize=2,\n",
    "        markeredgewidth=0.3 if 'Marker' not in label else 0,\n",
    "        label=label\n",
    "    )\n",
    "    for label, color in label_to_color.items()\n",
    "]\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "ax.legend(handles=handles, fontsize=4, loc='upper center', frameon=False, \n",
    "          bbox_to_anchor=(0.5, 0), ncol=2)\n",
    "\n",
    "\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "ax.spines[[\"top\", \"right\", \"left\", \"bottom\"]].set_linewidth(0.3)\n",
    "ax.set_xlim(-10,1810)\n",
    "ax.set_ylim(-10,1810)\n",
    "ax.set_aspect('equal')\n",
    "plt.title(\"MOD2 cells & marker transcripts\", fontsize=6)\n",
    "plt.tight_layout()  \n",
    "plt.savefig(fig_dir /\"MOD2marker.pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a6b486a",
   "metadata": {},
   "source": [
    "### VSI distribution within MOD subtypes\n",
    "VSI values of pixels within MOD1 and MOD2 cell boundaries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "216125b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from shapely.geometry import Polygon\n",
    "from matplotlib.path import Path as mPath\n",
    "\n",
    "def extract_cell_vsi(\n",
    "    boundary_df, \n",
    "    integrity,\n",
    "    strength,\n",
    "    integrity_size=1800\n",
    "):\n",
    "    \"\"\"\n",
    "    Extracts the cell integrity and strength arrays based on polygonal boundaries.\n",
    "    \"\"\"\n",
    "\n",
    "    cell_integrity = np.zeros((integrity_size, integrity_size))\n",
    "    cell_strength = np.zeros((integrity_size, integrity_size))\n",
    "\n",
    "    for idx, row in boundary_df.iterrows():\n",
    "        x_coords = np.array(row['boundaryX'])\n",
    "        y_coords = np.array(row['boundaryY'])\n",
    "\n",
    "        valid_mask = ~np.isnan(x_coords) & ~np.isnan(y_coords)\n",
    "        x_coords = x_coords[valid_mask]\n",
    "        y_coords = y_coords[valid_mask]\n",
    "\n",
    "        if len(x_coords) < 3:\n",
    "            continue\n",
    "\n",
    "        polygon = Polygon(zip(x_coords, y_coords))\n",
    "        if not polygon.is_valid:\n",
    "            continue\n",
    "\n",
    "        x_min, x_max = int(np.floor(polygon.bounds[0])), int(np.ceil(polygon.bounds[2]))\n",
    "        y_min, y_max = int(np.floor(polygon.bounds[1])), int(np.ceil(polygon.bounds[3]))\n",
    "\n",
    "        x_min, x_max = max(0, x_min), min(integrity_size, x_max)\n",
    "        y_min, y_max = max(0, y_min), min(integrity_size, y_max)\n",
    "\n",
    "        y_indices, x_indices = np.meshgrid(range(y_min, y_max), range(x_min, x_max), indexing='ij')\n",
    "        points = np.column_stack([x_indices.ravel(), y_indices.ravel()])\n",
    "\n",
    "        path = mPath(np.column_stack((x_coords, y_coords)))\n",
    "        epsilon = 0.5\n",
    "        mask = path.contains_points(points, radius=epsilon)\n",
    "        mask = mask.reshape(y_indices.shape)\n",
    "\n",
    "        subgrid_int = integrity[y_min:y_max, x_min:x_max]\n",
    "        subgrid_str = strength[y_min:y_max, x_min:x_max]\n",
    "\n",
    "        cell_integrity[y_min:y_max, x_min:x_max] = np.where(\n",
    "            mask, subgrid_int, cell_integrity[y_min:y_max, x_min:x_max]\n",
    "        )\n",
    "        cell_strength[y_min:y_max, x_min:x_max] = np.where(\n",
    "            mask, subgrid_str, cell_strength[y_min:y_max, x_min:x_max]\n",
    "        )\n",
    "\n",
    "    return cell_integrity, cell_strength"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd6cb6e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "MOD1_int, MOD1_str = extract_cell_vsi(MOD1_boundaries, signal_integrity, signal_strength)\n",
    "MOD2_int, MOD2_str = extract_cell_vsi(MOD2_boundaries, signal_integrity, signal_strength)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed7dee89",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_histogram(ax, cell_integrity, cell_strength, signal_threshold, cmap, label, ylim=(1e-1,32), title=None):\n",
    "    n, bins, patches = ax.hist(\n",
    "        cell_integrity[cell_strength > signal_threshold],\n",
    "        bins=50,\n",
    "        range=(0, 1),\n",
    "        density=True,\n",
    "        edgecolor='black',\n",
    "        linewidth=0.3,\n",
    "        alpha=0.8,\n",
    "        rasterized=True\n",
    "    )\n",
    "\n",
    "    for i, patch in enumerate(patches):\n",
    "        patch.set_facecolor(cmap(i / len(patches)))\n",
    "\n",
    "    ax.set_xlim(0, 1)\n",
    "    ax.set_ylim(ylim)\n",
    "    ax.set_yscale('log', base=2)\n",
    "    ax.set_ylabel(\"Density\", fontsize=5)\n",
    "    ax.set_xlabel(label, fontsize=5)\n",
    "    ax.spines[[\"top\", \"right\"]].set_visible(False)\n",
    "    ax.spines[['left','bottom']].set_linewidth(0.3)\n",
    "    ax.xaxis.set_tick_params(labelsize=5, width=0.3, length=1)\n",
    "    ax.yaxis.set_tick_params(labelsize=5, width=0.3, length=1)\n",
    "    ax.set_title(title, fontsize=6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9931d792",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1, figsize=(9*CM, 4*CM), dpi=600)\n",
    "\n",
    "plot_histogram(ax, \n",
    "               cell_integrity=MOD1_int, \n",
    "               cell_strength=MOD1_str, \n",
    "               signal_threshold=3, \n",
    "               cmap=BIH_CMAP, \n",
    "               label=\"Vertical Signal Integrity\", \n",
    "               ylim=(1e-1,64), \n",
    "               title=\"Vertical Signal Integrity within MOD1\")\n",
    "plt.savefig(fig_dir /\"VSImod1.pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e81bbaf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1, figsize=(9*CM, 4*CM), dpi=600)\n",
    "\n",
    "plot_histogram(ax, \n",
    "               cell_integrity=MOD2_int, \n",
    "               cell_strength=MOD2_str, \n",
    "               signal_threshold=3, \n",
    "               cmap=BIH_CMAP, \n",
    "               label=\"Vertical Signal Integrity\", \n",
    "               ylim=(1e-1,64), \n",
    "               title=\"Vertical Signal Integrity within MOD2\")\n",
    "plt.savefig(fig_dir /\"VSImod2.pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c7c5afa",
   "metadata": {},
   "source": [
    "### Marker expression in non-oligodendrocyte cells (scRNA-seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9978ec68",
   "metadata": {},
   "outputs": [],
   "source": [
    "noOD_sc_data = pd.concat([sc_DE, sc_cell_class], axis=1)\n",
    "noOD_sc_data = noOD_sc_data[~noOD_sc_data['Cell_class'].str.startswith(\"OD\")]\n",
    "\n",
    "noOD_sc_cell_class = noOD_sc_data['Cell_class']\n",
    "noOD_sc_data = noOD_sc_data.drop('Cell_class', axis=1)\n",
    "\n",
    "Gene_Group = pd.DataFrame([0,0,0,0,0,1,1,1,1,1,1,1,1,1], columns=['cluster'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f690bef1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.cluster.hierarchy import linkage, leaves_list\n",
    "import seaborn as sns\n",
    "\n",
    "HEATMAP_CMAP = sns.color_palette(\"RdYlBu_r\", as_cmap=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5fe361e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_annotate_heatmap(cluster_data, cluster_labels, gene_groups=None, zscore=True, cmap=HEATMAP_CMAP,\n",
    "                          cluster_text_y=-1.2, show_cluster=True,\n",
    "                          show_cluster_lines=True, DE_g_line=True):\n",
    "\n",
    "    cluster_data = cluster_data.copy()\n",
    "    cluster_data['Cell_class'] = cluster_labels\n",
    "\n",
    "    cluster_data = cluster_data.sort_values(by='Cell_class')\n",
    "    cell_class_col = cluster_data['Cell_class']\n",
    "\n",
    "    numeric_data = cluster_data.drop(columns=['Cell_class']).apply(pd.to_numeric, errors='coerce')\n",
    "    numeric_data = numeric_data.dropna(axis=1).loc[:, ~numeric_data.T.duplicated()]\n",
    "\n",
    "    cluster_data = pd.concat([cell_class_col, numeric_data], axis=1)\n",
    "    expression_data = cluster_data.drop('Cell_class', axis=1)\n",
    "    expression_data = expression_data.loc[:, ~expression_data.columns.duplicated()]\n",
    "\n",
    "    cluster_labels_sorted = cluster_data['Cell_class'].values\n",
    "    unique_labels = pd.Series(cluster_labels_sorted).unique()\n",
    "\n",
    "    if zscore:\n",
    "        expression_data = expression_data.apply(lambda x: (x - x.mean()) / x.std(), axis=0)\n",
    "        vmin, vmax = -3, 3\n",
    "    else:\n",
    "        vmin, vmax = 0, 5\n",
    "\n",
    "    new_order = []\n",
    "    for label in unique_labels:\n",
    "        indices = np.where(cluster_labels_sorted == label)[0]\n",
    "        subset = expression_data.iloc[indices, :]\n",
    "        if subset.shape[0] > 1:\n",
    "            linkage_matrix = linkage(subset, method='ward')\n",
    "            sorted_indices = indices[leaves_list(linkage_matrix)]\n",
    "        else:\n",
    "            sorted_indices = indices\n",
    "        new_order.extend(sorted_indices)\n",
    "\n",
    "    reordered_expression_data = expression_data.iloc[new_order, :]\n",
    "    reordered_cluster_labels = cluster_labels_sorted[new_order]\n",
    "\n",
    "    if gene_groups is not None:\n",
    "        new_gene_order = []\n",
    "        for label in sorted(set(gene_groups['cluster'])):\n",
    "            gene_indices = np.where(gene_groups == label)[0]\n",
    "            subset = reordered_expression_data.iloc[:, gene_indices]\n",
    "            if subset.shape[1] > 1:\n",
    "                linkage_matrix = linkage(subset.T, method='average')\n",
    "                sorted_gene_indices = gene_indices[leaves_list(linkage_matrix)]\n",
    "            else:\n",
    "                sorted_gene_indices = gene_indices\n",
    "            new_gene_order.extend(sorted_gene_indices)\n",
    "    else:\n",
    "        new_gene_order = leaves_list(linkage(reordered_expression_data.T, method='average'))\n",
    "\n",
    "    reordered_expression_data = reordered_expression_data.iloc[:, new_gene_order]\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(6*CM, 8*CM), dpi=600)\n",
    "    img = sns.heatmap(\n",
    "        reordered_expression_data,\n",
    "        vmin=vmin, vmax=vmax,\n",
    "        cmap=cmap,\n",
    "        xticklabels=True,\n",
    "        yticklabels=False,\n",
    "        cbar=True,\n",
    "        ax=ax,\n",
    "        cbar_kws={\"shrink\": 0.5}\n",
    "    )\n",
    "\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), fontsize=5, rotation=90)\n",
    "    ax.xaxis.set_tick_params(width=0.3, length=2)\n",
    "    ax.set_ylabel(\"\")\n",
    "    ax.set_xlabel(\"\")\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_linewidth(0.3)\n",
    "\n",
    "    cbar = img.collections[0].colorbar\n",
    "    cbar.ax.tick_params(labelsize=4, width=0.2, length=0.7)\n",
    "    for spine in cbar.ax.spines.values():\n",
    "        spine.set_linewidth(0.3)\n",
    "\n",
    "    cluster_boundaries = []\n",
    "    for label in unique_labels:\n",
    "        indices = np.where(reordered_cluster_labels == label)[0]\n",
    "        if len(indices) == 0:\n",
    "            continue\n",
    "        start, end = indices[0], indices[-1]\n",
    "        y_pos = (start + end) / 2\n",
    "        if show_cluster:\n",
    "            ax.text(cluster_text_y, y_pos, str(label), ha='center', va='center', rotation=0, fontsize=4)\n",
    "        cluster_boundaries.append(end)\n",
    "\n",
    "    if show_cluster_lines:\n",
    "        for boundary in cluster_boundaries[:-1]:\n",
    "            ax.axhline(y=boundary + 0.1, color='black', linestyle='--', linewidth=0.3)\n",
    "\n",
    "    if DE_g_line:\n",
    "        ax.vlines(x=5, ymin=ax.get_ylim()[0], ymax=ax.get_ylim()[1], color='black', linestyle='--', linewidth=0.3)\n",
    "\n",
    "    \n",
    "    plt.title(\"heatmap for markers in scRNA-seq data\", fontsize=6)\n",
    "    plt.savefig(fig_dir /\"heatmap.pdf\")\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f38260c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_annotate_heatmap(cluster_data = noOD_sc_data, cluster_labels=noOD_sc_cell_class, gene_groups=Gene_Group, cluster_text_y=-2.3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94d26dfa",
   "metadata": {},
   "source": [
    "### Overlapping ROIs\n",
    "Examples of MOD2 cells spatially overlapping with annotated excitatory and inhibitory neurons  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "633f3cf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "windowsize = 60\n",
    "window_size = windowsize / 2\n",
    "\n",
    "signal_threshold=3.0\n",
    "roi_scatter_kwargs = dict(marker=\".\", alpha=0.8, s=5)\n",
    "\n",
    "embedding = pseudocells.obsm[\"2D_UMAP\"]\n",
    "RGB = pseudocells.obsm[\"RGB\"]\n",
    "spatial = pseudocells.obsm[\"spatial\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b35a20d",
   "metadata": {},
   "source": [
    "#### UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21fdaaad",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(3*CM, 3*CM), dpi=600)\n",
    "ax.scatter(\n",
    "    embedding[:, 0],\n",
    "    embedding[:, 1],\n",
    "    c=RGB,\n",
    "    alpha=0.5,\n",
    "    marker=\".\",\n",
    "    edgecolors='none',\n",
    "    s=2,\n",
    ")\n",
    "ax.set(aspect='equal', adjustable='box')\n",
    "\n",
    "ax.set_xlim(-8,17)\n",
    "ax.set_ylim(-8,17)\n",
    "\n",
    "for spine in ax.spines.values():\n",
    "        spine.set_linewidth(0.3)\n",
    "\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "plt.title(\"UMAP\", fontsize=6)\n",
    "plt.tight_layout()\n",
    "plt.savefig(fig_dir /\"UMAP.pdf\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88316d62",
   "metadata": {},
   "source": [
    "#### Cell Map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09f927a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(3*CM, 3*CM), dpi=600)\n",
    "ax.scatter(\n",
    "    spatial[:, 0],\n",
    "    spatial[:, 1],\n",
    "    c=RGB,\n",
    "    alpha=0.8,\n",
    "    marker=\".\",\n",
    "    edgecolors='none',\n",
    "    s=2,\n",
    ")\n",
    "ax.set(aspect=\"equal\")\n",
    "\n",
    "x_center, y_center = doublets[\"x\", \"y\"].row(49)\n",
    "x0 = x_center - window_size\n",
    "y0 = y_center - window_size\n",
    "square = Rectangle(\n",
    "    (x0, y0),\n",
    "    width=windowsize,\n",
    "    height=windowsize,\n",
    "    linewidth=0.4,\n",
    "    fill=False,\n",
    "    edgecolor=\"k\",\n",
    ")\n",
    "ax.add_patch(square)\n",
    "\n",
    "x_center, y_center = doublets[\"x\", \"y\"].row(158)\n",
    "x0 = x_center - window_size\n",
    "y0 = y_center - window_size\n",
    "square = Rectangle(\n",
    "    (x0, y0),\n",
    "    width=windowsize,\n",
    "    height=windowsize,\n",
    "    linewidth=0.4,\n",
    "    fill=False,\n",
    "    edgecolor=\"k\",\n",
    ")\n",
    "ax.add_patch(square)\n",
    "\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "for spine in ax.spines.values():\n",
    "        spine.set_linewidth(0.3)\n",
    "ax.set_xlim(-30,1830)\n",
    "ax.set_ylim(-30,1830)\n",
    "ax.set_aspect('equal')\n",
    "plt.title(\"cell type\", fontsize=6)\n",
    "plt.tight_layout()\n",
    "plt.savefig(fig_dir /\"CellTypeOvrlpy.pdf\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d1a8eca",
   "metadata": {},
   "source": [
    "#### Doublet1: Inhibitory\n",
    "\n",
    "doublet_to_show = 49  \n",
    "x,y = 292, 924"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "335db986",
   "metadata": {},
   "outputs": [],
   "source": [
    "doublet_to_show = 49\n",
    "x, y = doublets[\"x\", \"y\"].row(doublet_to_show)\n",
    "roi = ((x - window_size, x + window_size), (y - window_size, y + window_size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fba25a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "ROI_in  = dataset.transcripts.filter(\n",
    "            pl.col(\"x\").is_between(x - window_size, x + window_size)\n",
    "            & pl.col(\"y\").is_between(y - window_size, y + window_size)\n",
    "        ).clone().sort(\"z\")\n",
    "\n",
    "_, embedding_color = dataset.transform_transcripts(ROI_in)\n",
    "ROI_in = ROI_in.with_columns(RGB=embedding_color)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "027c8d49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# VSI map\n",
    "fig, ax = plt.subplots(figsize=(3*CM, 3*CM), dpi=600)\n",
    "\n",
    "ax.set_facecolor(\"black\")\n",
    "img = ax.imshow(\n",
    "    signal_integrity,\n",
    "    cmap=BIH_CMAP,\n",
    "    alpha=((signal_strength / signal_threshold).clip(0, 1) ** 2),\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    ")\n",
    "ax.invert_yaxis()\n",
    "ax.set(xlim=roi[0], ylim=roi[1])\n",
    "\n",
    "cbar = fig.colorbar(img, ax=ax, shrink=0.8)\n",
    "cbar.ax.tick_params(labelsize=6, width=0.3)\n",
    "for spine in cbar.ax.spines.values():\n",
    "    spine.set_linewidth(0.3)\n",
    "\n",
    "_plot_scalebar(ax, dx=1, units=\"um\", location=\"lower left\", length_fraction=0.4, fontsize=5, color=\"white\")\n",
    "\n",
    "ax.set(xticks=[], yticks=[])\n",
    "ax.set_title(\"vertical signal integrity\", fontsize=6)\n",
    "ax.set_aspect('equal')\n",
    "\n",
    "plt.savefig(fig_dir / \"INvsi.pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83d1e779",
   "metadata": {},
   "outputs": [],
   "source": [
    "roi_top = ROI_in.filter(pl.col(\"z\") > pl.col(\"z_center\"))\n",
    "\n",
    "(x_min, x_max), (y_min, y_max) = roi\n",
    "fig, ax = plt.subplots(figsize=(3*CM, 3*CM), dpi=600)\n",
    "ax.scatter(\n",
    "    roi_top[\"x\"],\n",
    "    roi_top[\"y\"],\n",
    "    c=roi_top[\"RGB\"].to_numpy(),\n",
    "    **roi_scatter_kwargs,\n",
    "    edgecolors='none'\n",
    ")\n",
    "ax.set(xlim=(x_min, x_max), ylim=(y_min, y_max))\n",
    "\n",
    "filtered_df = boundaries_df[\n",
    "    (boundaries_df['x'] >= x_min) & (boundaries_df['x'] <= x_max) &\n",
    "    (boundaries_df['y'] >= y_min) & (boundaries_df['y'] <= y_max)\n",
    "]\n",
    "\n",
    "ax.spines[[\"top\", \"right\"]].set_visible(True)\n",
    "for _, row in filtered_df.iterrows():\n",
    "    ax.plot(row['boundaryX'], row['boundaryY'], c='#2C2C2C', linewidth=0.7)\n",
    "\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "ax.set_aspect('equal')\n",
    "\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_linewidth(0.3)\n",
    "\n",
    "plt.title(\"top\", fontsize=6)\n",
    "plt.tight_layout()\n",
    "plt.savefig(fig_dir / \"INtop.pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbd709fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "roi_bottom = ROI_in.filter(pl.col(\"z\") < pl.col(\"z_center\"))[::-1]\n",
    "(x_min, x_max), (y_min, y_max) = roi\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3*CM, 3*CM), dpi=600)\n",
    "ax.scatter(\n",
    "    roi_bottom[\"x\"],\n",
    "    roi_bottom[\"y\"],\n",
    "    c=roi_bottom[\"RGB\"].to_numpy(),\n",
    "    **roi_scatter_kwargs,\n",
    "    edgecolors='none',\n",
    ")\n",
    "ax.set(xlim=(x_min, x_max), ylim=(y_min, y_max))\n",
    "\n",
    "filtered_df = boundaries_df[\n",
    "    (boundaries_df['x'] >= x_min) & (boundaries_df['x'] <= x_max) &\n",
    "    (boundaries_df['y'] >= y_min) & (boundaries_df['y'] <= y_max)\n",
    "]\n",
    "\n",
    "for _, row in filtered_df.iterrows():\n",
    "    ax.plot(row['boundaryX'], row['boundaryY'], c='#2C2C2C', linewidth=0.7)\n",
    "\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "ax.set_aspect('equal')\n",
    "\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_linewidth(0.3)\n",
    "\n",
    "plt.title(\"bottom\", fontsize=6)\n",
    "plt.tight_layout()  \n",
    "plt.savefig(fig_dir /\"INbottom.pdf\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7807e7d9",
   "metadata": {},
   "source": [
    "#### Doublet2: Excitatory\n",
    "\n",
    "doublet_to_show = 158  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "add5eee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "doublet_to_show = 158\n",
    "x, y = doublets[\"x\", \"y\"].row(doublet_to_show)\n",
    "roi = ((x - window_size, x + window_size), (y - window_size, y + window_size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0147a9ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "ROI_ex  = dataset.transcripts.filter(\n",
    "            pl.col(\"x\").is_between(x - window_size, x + window_size)\n",
    "            & pl.col(\"y\").is_between(y - window_size, y + window_size)\n",
    "        ).clone().sort(\"z\")\n",
    "\n",
    "_, embedding_color = dataset.transform_transcripts(ROI_ex)\n",
    "ROI_ex = ROI_ex.with_columns(RGB=embedding_color)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76a1865a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# VSI map\n",
    "fig, ax = plt.subplots(figsize=(3*CM, 3*CM), dpi=600)\n",
    "\n",
    "ax.set_facecolor(\"black\")\n",
    "img = ax.imshow(\n",
    "    signal_integrity,\n",
    "    cmap=BIH_CMAP,\n",
    "    alpha=((signal_strength / signal_threshold).clip(0, 1) ** 2),\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    ")\n",
    "ax.invert_yaxis()\n",
    "ax.set(xlim=roi[0], ylim=roi[1])\n",
    "\n",
    "cbar = fig.colorbar(img, ax=ax, shrink=0.8)\n",
    "cbar.ax.tick_params(labelsize=6, width=0.3)\n",
    "for spine in cbar.ax.spines.values():\n",
    "    spine.set_linewidth(0.3)\n",
    "\n",
    "ax.set(xticks=[], yticks=[])\n",
    "ax.set_title(\"vertical signal integrity\", fontsize=6)\n",
    "ax.set_aspect('equal')\n",
    "\n",
    "plt.savefig(fig_dir / \"EXvsi.pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "711c2ca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "roi_top = ROI_ex.filter(pl.col(\"z\") > pl.col(\"z_center\"))\n",
    "(x_min, x_max), (y_min, y_max) = roi\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3*CM, 3*CM), dpi=600)\n",
    "ax.scatter(\n",
    "    roi_top[\"x\"],\n",
    "    roi_top[\"y\"],\n",
    "    c=roi_top[\"RGB\"].to_numpy(),\n",
    "    **roi_scatter_kwargs,\n",
    "    edgecolors='none',\n",
    "    rasterized=True\n",
    ")\n",
    "\n",
    "filtered_df = boundaries_df[\n",
    "    (boundaries_df['x'] >= x_min) & (boundaries_df['x'] <= x_max) &\n",
    "    (boundaries_df['y'] >= y_min) & (boundaries_df['y'] <= y_max)\n",
    "]\n",
    "\n",
    "ax.spines[[\"top\", \"right\"]].set_visible(True)\n",
    "for _, row in filtered_df.iterrows():\n",
    "    ax.plot(row['boundaryX'], row['boundaryY'], c='#2C2C2C', linewidth=0.7)\n",
    "\n",
    "ax.set(xlim=(x_min, x_max), ylim=(y_min, y_max))\n",
    "ax.set(xticks=[], yticks=[])\n",
    "ax.set_aspect('equal')\n",
    "\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_linewidth(0.3)\n",
    "\n",
    "plt.title(\"top\", fontsize=6)\n",
    "plt.tight_layout()\n",
    "plt.savefig(fig_dir / \"EXtop.pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a1c7aea",
   "metadata": {},
   "outputs": [],
   "source": [
    "roi_bottom = ROI_ex.filter(pl.col(\"z\") < pl.col(\"z_center\"))[::-1]\n",
    "(x_min, x_max), (y_min, y_max) = roi\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3*CM, 3*CM), dpi=600)\n",
    "ax.scatter(\n",
    "    roi_bottom[\"x\"],\n",
    "    roi_bottom[\"y\"],\n",
    "    c=roi_bottom[\"RGB\"].to_numpy(),\n",
    "    **roi_scatter_kwargs,\n",
    "    edgecolors='none',\n",
    ")\n",
    "\n",
    "filtered_df = boundaries_df[\n",
    "    (boundaries_df['x'] >= x_min) & (boundaries_df['x'] <= x_max) &\n",
    "    (boundaries_df['y'] >= y_min) & (boundaries_df['y'] <= y_max)\n",
    "]\n",
    "\n",
    "ax.spines[[\"top\", \"right\"]].set_visible(True)\n",
    "for _, row in filtered_df.iterrows():\n",
    "    ax.plot(row['boundaryX'], row['boundaryY'], c='#2C2C2C', linewidth=0.7)\n",
    "\n",
    "ax.set(xlim=(x_min, x_max), ylim=(y_min, y_max))\n",
    "ax.set(xticks=[], yticks=[])\n",
    "ax.set_aspect('equal')\n",
    "\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_linewidth(0.3)\n",
    "\n",
    "plt.title(\"bottom\", fontsize=6)\n",
    "plt.tight_layout()\n",
    "plt.savefig(fig_dir / \"EXbottom.pdf\")\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ovrlpyrevision",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
